### PR TITLE
[pinmux] Add strap sampling and TAP qualification logic

### PR DESCRIPTION
**Pull request by msfschaffner**
This implements the strap sampling logic and the TAP qualification with life cycle signals.

Signed-off-by: Michael Schaffner <msf@opentitan.org>

### CODE REVIEW

**Comment by msfschaffner** [comment]
Related feature tracker: https://github.com/lowRISC/opentitan/issues/5221

File: hw/ip/pinmux/rtl/pinmux_strap_sampling.sv [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    .rsp_i(dft_jtag_i),
+    .rsp_o(dft_jtag_rsp)
+  );
+
+  ////////////////
+  // Assertions //
+  ////////////////
+
+  // The strap sampling enable input shall be pulsed high exactly once after cold boot.
+  `ASSERT(PwrMgrStrapSampleOnce_A, strap_en_i |=> strong(!strap_en_i [*]))
```
Comment (cindychip) [inline hw/ip/pinmux/rtl/pinmux_strap_sampling.sv:216]: nice syntax! 
I wonder if keyword `strong` is needed here.
I think ` strap_en_i |=> !strap_en_i [*]`, once found strap_en_i, from next cycle, strap_en_i has to stay low until next reset.
Also a quick question: does the `strap_en_i` pulse only set high for one cycle?

**Review by cindychip — COMMENTED** [review]
Thanks for tagging me on the PR michael! I mainly read through and focused on the assertions. LGTM but would be good to have some RTL engineers approve it.

File: hw/ip/pinmux/rtl/pinmux_strap_sampling.sv (hunk 1 above)
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux_strap_sampling.sv:216]: Hey @cindychip, you are probably right that we do not need the `strong` modification here.
Let me remove that.
RE your question regarding the enable signal: yes, it is expected that this is a single cycle pulse.

File: hw/ip/pinmux/rtl/pinmux_strap_sampling.sv [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    .rsp_i(dft_jtag_i),
+    .rsp_o(dft_jtag_rsp)
+  );
+
+  ////////////////
+  // Assertions //
+  ////////////////
+
+  // The strap sampling enable input shall be pulsed high exactly once after cold boot.
+  `ASSERT(PwrMgrStrapSampleOnce_A, strap_en_i |=> !strap_en_i [*])
```
Comment (cindychip) [inline hw/ip/pinmux/rtl/pinmux_strap_sampling.sv:216]: sorry i forget to mention, you need to add ##0 to pass:
```
`ASSERT(PwrMgrStrapSampleOnce_A, strap_en_i |=> ##0  !strap_en_i [*])
```
reference here: https://verificationacademy.com/forums/systemverilog/regarding-assertion-checking

File: hw/top_earlgrey/rtl/autogen/top_earlgrey.sv [hunk 3]
```diff
@@ -441,6 +441,8 @@ module top_earlgrey #(
   keymgr_pkg::kmac_data_req_t       keymgr_kmac_data_req;
   keymgr_pkg::kmac_data_rsp_t       keymgr_kmac_data_rsp;
   logic [3:0] clkmgr_aon_idle;
+  jtag_pkg::jtag_req_t       pinmux_aon_lc_jtag_req;
+  jtag_pkg::jtag_rsp_t       pinmux_aon_lc_jtag_rsp;
```
Comment (cindychip) [inline hw/top_earlgrey/rtl/autogen/top_earlgrey.sv:445]: Looks like there is a multi-driver issue:
```
Error-[ICSD] Illegal combination of drivers
../src/lowrisc_systems_top_earlgrey_0.1/rtl/autogen/top_earlgrey.sv, 445
  Illegal combination of structural drivers.
  Variable "pinmux_aon_lc_jtag_rsp" is driven by multiple structural drivers.
  This variable is declared at
```

File: hw/ip/pinmux/rtl/pinmux_strap_sampling.sv (hunk 2 above)
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux_strap_sampling.sv]: Thanks @cindychip!

File: hw/top_earlgrey/rtl/autogen/top_earlgrey.sv (hunk 3 above)
Comment (msfschaffner) [reply hw/top_earlgrey/rtl/autogen/top_earlgrey.sv:445]: Yeah looks like I accidentally reversed the `req`/`rec` directions in the `pinmux.hjson` file, causing topgen to incorrectly connect these wires...

File: hw/ip/pinmux/rtl/pinmux.sv [hunk 4]
```diff
@@ -19,12 +19,22 @@ module pinmux
...
+  input  lc_ctrl_pkg::lc_tx_t      lc_hw_debug_en_i,
+  // Sampled values for DFT straps
   output dft_strap_test_req_t      dft_strap_test_o,
+  // Qualified JTAG signals for TAPs
+  output jtag_pkg::jtag_req_t      lc_jtag_o,
+  input  jtag_pkg::jtag_rsp_t      lc_jtag_i,
+  output jtag_pkg::jtag_req_t      rv_jtag_o,
+  input  jtag_pkg::jtag_rsp_t      rv_jtag_i,
+  output jtag_pkg::jtag_req_t      dft_jtag_o,
+  input  jtag_pkg::jtag_rsp_t      dft_jtag_i,
```
Comment (tjaychen) [inline hw/ip/pinmux/rtl/pinmux.sv:37]: quick question, since the flash technically also has a jtag connection to pinmux (however it is muxed and nothing special), should i just do that as part of normal cio_'s?

File: hw/ip/pinmux/rtl/pinmux.sv [hunk 5]
```diff
@@ -19,12 +19,22 @@ module pinmux
...
   output logic                     aon_wkup_req_o,
   output logic                     usb_wkup_req_o,
-  // Sleep enable, running on clk_i
+  // Sleep enable and strap sample enable
+  // from pwrmgr, running on clk_i
   input                            sleep_en_i,
-  // Strap sample request
-  input  lc_strap_req_t            lc_pinmux_strap_i,
-  output lc_strap_rsp_t            lc_pinmux_strap_o,
+  input                            strap_en_i,
```
Comment (tjaychen) [inline hw/ip/pinmux/rtl/pinmux.sv:25]: would you mind attaching to this #5198 also? that way i won't forget to add it.

File: hw/ip/pinmux/rtl/pinmux_strap_sampling.sv [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    end
+
+    // In case DFT is enabled, and in case
+    // the TAP straps where set to 11 (DFT) upon
+    // the first sampling event, we continue sampling
+    // all straps until system reset. This is used
+    // during the DFT-enabled life cycle states only.
+    if (lc_dft_en[0] == lc_ctrl_pkg::On) begin
+      if (strap_en_q &&
+          tap_strap_t'(tap_strap_q) == DftTapSel) begin
```
Comment (tjaychen) [inline hw/ip/pinmux/rtl/pinmux_strap_sampling.sv:118]: does this pre-condition `DftTapSel` to be the first one? Because I'm not sure that would always be true.  I thought it was more that as long as we are in `dft_en` mode, we allow this to continually change? 

Let me know if I remembered incorrectly (this shouldn't block the PR, we can always change it later if needed)

File: hw/ip/pinmux/rtl/pinmux_strap_sampling.sv (hunk 1 above)
Comment (tjaychen) [reply hw/ip/pinmux/rtl/pinmux_strap_sampling.sv]: dumb question from me...  what would the `strong` modification imply here..?

**Review by tjaychen — APPROVED** [review]
lgtm with a few clarifying questions. 
They don't need to be addressed in this PR.

File: hw/ip/pinmux/rtl/pinmux.sv (hunk 4 above)
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux.sv:37]: Hm... let me see... Afaik, the flash JTAG has a different LC qualification signal (the NVM debug one, right)? So we can't daisy chain the DFT TAP with the flash one?

We could technically add a strap pin bit to allow selection of other TAPs here as well...

Otherwise, if it is fine from a test-flow perspective that the JTAG is functionally muxed in the pinmux ( (and therefore requires a pinmux configuration), then we could go for that, yes.

File: hw/ip/pinmux/rtl/pinmux.sv (hunk 5 above)
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux.sv:25]: oh yeah, let me do that.

File: hw/ip/pinmux/rtl/pinmux.sv (hunk 4 above)
Comment (tjaychen) [reply hw/ip/pinmux/rtl/pinmux.sv:37]: yeah it's a different signal, and I do the controls for that directly in `flash_phy` already, so we don't have to expose anything to pinmux.  Maybe that's better then, I just stick with the vanilla pinmux flow and we don't have to do anything extra here.

File: hw/ip/pinmux/rtl/pinmux_strap_sampling.sv (hunk 6 above)
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux_strap_sampling.sv:118]: I quickly checked our notes in Monorail:
```
More precisely, if we are in TEST and the strap sampling circuitry determines at POR that the strap pins are non-zero, we shall keep on sampling the strap pins continuously to enable switching between different TAPs during one test run.
```
So I guess the preconditioning should be on straps != 0. let me fix that.

File: hw/ip/pinmux/rtl/pinmux.sv (hunk 4 above)
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux.sv:37]: ok SGTM!

**Comment by tjaychen** [comment]
ah nice! monorail to the rescue!

File: hw/ip/pinmux/rtl/pinmux_strap_sampling.sv (hunk 6 above)
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux_strap_sampling.sv:118]: fixed

File: hw/ip/pinmux/rtl/pinmux_strap_sampling.sv (hunk 1 above)
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux_strap_sampling.sv]: it ensures that the sequential property holds only if the sequence has matched (e.g., you can't have sequences that have started evaluation but have not matched yet at the end of a simulation).

this is needed for things like testing that eventually, an FSM returns to a certain state (especially in FPV, although simulation can sometimes also check this).

I placed it here although it was not necessary in this context, since the sequence always matches after pulse deassertion.
